technology_and_tooling/testing/testable_code_fixtures.md

```diff
@@ -0,0 +1,799 @@
+                            [0, 0],
+                            [0, 0]])
+        test_result = np.array([0, 0])
+
+        # Need to use Numpy testing functions to compare arrays
+        npt.assert_array_equal(trial_instance.daily_mean(), test_result)
+
+
+    def test_daily_mean_integers(self, trial_instance):
+        """Test that mean function works for an array of positive integers."""
+
+        trial_instance.data = np.array([[1, 2],
+                            [3, 4],
+                            [5, 6]])
+        test_result = np.array([3, 4])
+
+        # Need to use Numpy testing functions to compare arrays
+        npt.assert_array_equal(trial_instance.daily_mean(), test_result)
+
+
+    @pytest.mark.parametrize(
+        "test, expected",
+        [
+            ([ [0, 0, 0], [0, 0, 0], [0, 0, 0] ], [0, 0, 0]),
+            ([ [4, 2, 5], [1, 6, 2], [4, 1, 9] ], [4, 6, 9]),
+            ([ [4, -2, 5], [1, -6, 2], [-4, -1, 9] ], [4, -1, 9]),
+        ])
+    def test_daily_max(self, test, expected, trial_instance):
+        """Test max function works for zeroes, positive integers, mix of positive/negative integers."""
+        trial_instance.data = np.array(test)
+        npt.assert_array_equal(trial_instance.daily_max(), np.array(expected))
+
+    ...
+~~~
+
+:::
+::::
+
+In our tests for the `Trial` class, we have to initialise the class using a CSV file in order to create an instance, even if we do not use that particular data in our tests.  How can we simplify this? One thing that can be changed is the `__init__` method, if we just needed the data as an argument, rather than the path to a CSV file, that would make testing easier. After this change, a separate method is going to be needed to allow creating a `Trial` from a CSV filepath, this can be achieved using a class method. 
+
+
+::::challenge{id=load-from-csv title="Refactor the `Trial` class."}
+
+As described above, refactor the `--init__` method of the `Trial` class to take `data` and `id` as arguments. Then create a class method called `from_csv` that accepts a filepath to a CSV file and returns an instance of the `Trial` class with the data loaded.
+
+:::solution
+Here is the first section of our adjusted object code:
+~~~python
+class Trial:
+    def __init__(self, data, id):
+        self.data = data
+        self.id = id
+
+   @classmethod
+    def from_csv(cls, filename, id):
+        """
+        Class method to create a Trial instance from data in a CSV file.
+
+        Parameters:
+        filename (str): The file path of the CSV file to read.
+        id (str): The id to assign to the Trial instance.
+
+        Returns:
+        Trial: A Trial instance with the data and id from the CSV file.
+        """
+        data = cls.load_csv(filename)
+        return cls(data, id)
+
+    @staticmethod
+    def load_csv(filename):
+        """Load a Numpy array from a CSV
+
+        :param filename: Filename of CSV to load
```

Comment:
your `from_csv` and `load_csv` are using different docstring formats, should make these consistent

Reply:
Have fixed this now, thanks!